### PR TITLE
[SPARK-35253][SQL][BUILD] Bump up the janino version to v3.1.4

### PR DESCRIPTION
**Pull request by maropu**
### What changes were proposed in this pull request?

This PR proposes to bump up the janino version from 3.0.16 to v3.1.4.
The major changes of this upgrade are as follows:
 - Fixed issue #131: Janino 3.1.2 is 10x slower than 3.0.11: The Compiler's IClassLoader was initialized way too eagerly, thus lots of classes were loaded from the class path, which is very slow.
 - Improved the encoding of stack map frames according to JVMS11 4.7.4: Previously, only "full_frame"s were generated.
 - Fixed issue #107: Janino requires "org.codehaus.commons.compiler.io", but commons-compiler does not export this package
 - Fixed the promotion of the array access index expression (see JLS7 15.13 Array Access Expressions).

For all the changes, please see the change log: http://janino-compiler.github.io/janino/changelog.html

NOTE1: I've checked that there is no obvious performance regression. For all the data, see a link: https://docs.google.com/spreadsheets/d/1srxT9CioGQg1fLKM3Uo8z1sTzgCsMj4pg6JzpdcG6VU/edit?usp=sharing

NOTE2: We upgraded janino to 3.1.2 (#27860) once before, but the commit had been reverted in #29495 because of the correctness issue. Recently, #32374 had checked if Spark could land on v3.1.3 or not, but a new bug was found there. These known issues has been fixed in v3.1.4 by following PRs:
 - janino-compiler/janino#145
 - janino-compiler/janino#146

### Why are the changes needed?

janino v3.0.X  is no longer maintained.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

GA passed.

### CODE REVIEW

**Comment by maropu** [comment]
I've fixed some issues in janino and the latest janino (w/ these fixes) has been released hours ago.
 - https://github.com/janino-compiler/janino/pull/145
 - https://github.com/janino-compiler/janino/pull/146

I will check if the existing tests can pass and there is no TPCDS performance regression.

**Comment by dongjoon-hyun** [comment]
Thank you, @maropu . Yes, it will be great if we can succeed to upgrade at this time.

**Comment by kiszk** [comment]
It would be great if we could check some compilation time.
Janino may have degradation regarding the compilation time like https://github.com/apache/spark/pull/29495#issuecomment-683961394

**Comment by viirya** [comment]
Looks like all tests are passed!

**Comment by HyukjinKwon** [comment]
cc @rednaxelafx too FYI

**Comment by maropu** [comment]
okay, as far as I checked, I couldn't find any critical problem. I think it is good timing to re-start the discussion to decide whether we upgrade janino or not for the next release (v3.2.0). WDYT, guys? @gatorsmile @cloud-fan @HyukjinKwon @dongjoon-hyun @viirya @kiszk @rednaxelafx @HeartSaVioR @gengliangwang @tgravescs @LuciferYang (I've listed up the members who participated in the previous discussions)

**Comment by tgravescs** [comment]
3.2.0 sounds fine to me. From the description it seems like the issue in https://issues.apache.org/jira/browse/SPARK-32640 was  fixed with https://github.com/janino-compiler/janino/pull/145 so I'm good with it.  It might be nice to add a test like SPARK-32640 if we don't have one yet.

**Comment by viirya** [comment]
There was a test added in SPARK-32640. So looks it is fixed now as all tests are passed. If no regression found, looks okay.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM from my side.

**Comment by rednaxelafx** [comment]
I'll do some experiments in my environment as well. Thank you very much for all the efforts driving the Janino upgrade, @maropu san!

**Review by gengliangwang — APPROVED** [review]
+1. I just tried it locally and the bug described in https://issues.apache.org/jira/browse/SPARK-32640 is fixed.
Thanks for the work.

**Review by viirya — APPROVED** [review]

**Review by HyukjinKwon — APPROVED** [review]

**Comment by srowen** [comment]
Yes I'd update to get all those fixes if there's no reason to expect a problem now.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala [hunk 1]
```diff
@@ -1434,9 +1435,10 @@ object CodeGenerator extends Logging {
   private def updateAndGetCompilationStats(evaluator: ClassBodyEvaluator): ByteCodeStats = {
     // First retrieve the generated classes.
     val classes = {
-      val resultField = classOf[SimpleCompiler].getDeclaredField("result")
-      resultField.setAccessible(true)
-      val loader = resultField.get(evaluator).asInstanceOf[ByteArrayClassLoader]
+      val scField = classOf[ClassBodyEvaluator].getDeclaredField("sc")
```
[the remark this answers is not in the document]
Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala:1438]: Is the data the totally same with `ByteArrayClassLoader.classes`? The suggestion is a trivial update though, IMO this PR just focuses on the janino upgrade. Could you make a follow-up PR for it after this merged?

**Review by LuciferYang — APPROVED** [review]

**Review by cloud-fan — APPROVED** [review]

**Comment by maropu** [comment]
Thank you for all the reviewers~

**Comment by maropu** [comment]
@cloud-fan @HyukjinKwon As you know, the janino community is pretty inactive now. So, I'm wondering if it would be better to revert this commit from branch-3.2 before the QA period. WDYT? cc: @dongjoon-hyun @viirya

**Comment by LuciferYang** [comment]
@maropu If we need to revert this commit, remember to revert SPARK-35398 first

**Comment by srowen** [comment]
What's the need to revert it?

**Comment by cloud-fan** [comment]
Because the new janino version has bugs that can fail certain SQL queries, which is a regression.

**Comment by srowen** [comment]
Sounds good, revert it

**Comment by cloud-fan** [comment]
I'll open a revert PR soon. If janino can have a new release with the bug fixed before rc1, we can revert my revert.

**Comment by dongjoon-hyun** [comment]
+1 for reverting, too.

**Comment by gengliangwang** [comment]
+1 for reverting. Thank you @maropu @cloud-fan

**Comment by viirya** [comment]
+1 for reverting it. Thanks!

**Comment by cloud-fan** [comment]
The PR is out: https://github.com/apache/spark/pull/33302

**Pull request closed without merging**